pkg/shares/share_builder.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	return &b
+}
+
+func (b *Builder) AvailableBytes() int {
+	return appconsts.ShareSize - len(b.rawShareData)
+}
+
+func (b *Builder) ImportRawShare(rawBytes []byte) *Builder {
+	b.rawShareData = rawBytes
+	return b
+}
+
+func (b *Builder) AddData(rawData []byte) (rawDataLeftOver []byte) {
+	// find the len left in the pending share
+	pendingLeft := appconsts.ShareSize - len(b.rawShareData)
+
+	// if we can simply add the tx to the share without creating a new
+	// pending share, do so and return
+	if len(rawData) <= pendingLeft {
+		b.rawShareData = append(b.rawShareData, rawData...)
+		return nil
+	}
+
+	// if we can only add a portion of the rawData to the pending share,
+	// then we add it and add the pending share to the finalized shares.
+	chunk := rawData[:pendingLeft]
+	b.rawShareData = append(b.rawShareData, chunk...)
+
+	// We need to finish this share and start a new one
+	// so we return the leftover to be written into a new share
+	return rawData[pendingLeft:]
+}
+
+func (b *Builder) Build() (*Share, error) {
+	return newShare(b.rawShareData)
+}
+
+// IsEmptyShare returns true if no data has been written to the share
+func (b *Builder) IsEmptyShare() bool {
+	expectedLen := appconsts.NamespaceSize + appconsts.ShareInfoBytes
+	if b.isCompactShare {
+		expectedLen += appconsts.CompactShareReservedBytes
+	}
+	if b.isFirstShare {
+		expectedLen += appconsts.SequenceLenBytes
+	}
+	return len(b.rawShareData) == expectedLen
+}
+
+func (b *Builder) ZeroPadIfNecessary() (bytesOfPadding int) {
+	b.rawShareData, bytesOfPadding = zeroPadIfNecessary(b.rawShareData, appconsts.ShareSize)
+	return bytesOfPadding
+}
+
+// isEmptyReservedBytes returns true if the reserved bytes are empty.
+func (b *Builder) isEmptyReservedBytes() bool {
+	indexOfReservedBytes := b.indexOfReservedBytes()
+	reservedBytes, err := ParseReservedBytes(b.rawShareData[indexOfReservedBytes : indexOfReservedBytes+appconsts.CompactShareReservedBytes])
+	if err != nil {
+		panic(err)
+	}
+	return reservedBytes == 0
+}
+
+// indexOfReservedBytes returns the index of the reserved bytes in the share.
+func (b *Builder) indexOfReservedBytes() int {
+	if b.isFirstShare {
+		// if the share is the first share, the reserved bytes follow the namespace, info byte, and sequence length
+		return appconsts.NamespaceSize + appconsts.ShareInfoBytes + appconsts.SequenceLenBytes
+	}
+	// if the share is not the first share, the reserved bytes follow the namespace and info byte
+	return appconsts.NamespaceSize + appconsts.ShareInfoBytes
+}
+
+// indexOfInfoBytes returns the index of the InfoBytes.
+func (b *Builder) indexOfInfoBytes() int {
+	// the info byte is immediately after the namespace
+	return appconsts.NamespaceSize
+}
+
+// MaybeWriteReservedBytes will be a no-op if the reserved bytes
+// have already been populated. If the reserved bytes are empty, it will write
+// the location of the next unit of data to the reserved bytes.
+func (b *Builder) MaybeWriteReservedBytes() error {
```

Comment:
if we're going to throw an error and panic on that error, we should document when we can error.